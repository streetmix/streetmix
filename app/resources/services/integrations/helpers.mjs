import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
