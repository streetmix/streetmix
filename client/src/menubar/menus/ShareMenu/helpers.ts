import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
