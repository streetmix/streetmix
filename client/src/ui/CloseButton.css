button.close {
  --close-button-margin: 2px;
  --close-button-size: 34px;
  --close-icon-color: var(--color-turquoise-800);
  --close-icon-color-hover: var(--color-turquoise-700);

  /* Position in dialog boxes */
  position: absolute;
  right: var(--close-button-margin);
  top: var(--close-button-margin);

  /* Touch-friendly size */
  width: var(--close-button-size);
  height: var(--close-button-size);

  /* Appearance */
  display: flex;
  justify-content: center;
  align-items: center;
  appearance: none;
  padding: 0;
  border: 0;
  border-radius: var(--border-radius-pill);
  color: var(--close-icon-color);
  cursor: pointer;
  user-select: none;
  background-color: transparent;

  /* Mirror position in rtl */
  [dir="rtl"] & {
    right: auto;
    left: var(--close-button-margin);
  }

  &:hover {
    background-color: rgb(255 255 255 / 50%);
    color: var(--close-icon-color-hover);
  }

  &:active {
    background-color: rgb(255 255 255 / 75%);
  }

  /* !important overrides higher-level specificity on font awesome
     icons that normally exist on buttons. Apply color to the parent
     element (.close) and the icon will take on this color. */
  svg {
    color: inherit !important;
  }

  &:hover svg {
    color: inherit !important;
  }
}
