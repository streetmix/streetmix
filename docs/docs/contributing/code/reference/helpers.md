---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
