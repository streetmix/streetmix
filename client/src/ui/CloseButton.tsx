import React, { memo } from 'react'
import { useIntl } from 'react-intl'

import Icon from './Icon'
import './CloseButton.scss'

interface CloseButtonProps {
  title?: string
  className?: string
  onClick: (event: React.MouseEvent) => void
}

function CloseButton ({
  title,
  className = 'close',
  onClick,
  ...restProps
}: CloseButtonProps): React.ReactElement {
  const defaultTitle = useIntl().formatMessage({
    id: 'btn.dismiss',
    defaultMessage: 'Dismiss'
  })

  return (
    <button
      className={className}
      onClick={onClick}
      title={title ?? defaultTitle}
      {...restProps}
    >
      <Icon name="close" />
    </button>
  )
}

export default memo(CloseButton)
