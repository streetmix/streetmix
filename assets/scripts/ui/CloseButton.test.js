/* eslint-env jest */
import React from 'react'
import { userEvent } from '@testing-library/user-event'
import { render } from '../../../test/helpers/render'
import CloseButton from './CloseButton'

describe('CloseButton', () => {
  it('renders snapshot', () => {
    const { asFragment } = render(<CloseButton onClick={jest.fn()} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders with custom title, class name, and other attributes', () => {
    const { asFragment } = render(
      <CloseButton
        onClick={jest.fn()}
        title="foofoo"
        className="my-class"
        disabled={true}
        hidden={true}
      />
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should call onClick function when button is clicked', async () => {
    const onClick = jest.fn()
    const { getByTitle } = render(<CloseButton onClick={onClick} title="foo" />)
    await userEvent.click(getByTitle('foo'))
    expect(onClick).toHaveBeenCalledTimes(1)
  })
})
