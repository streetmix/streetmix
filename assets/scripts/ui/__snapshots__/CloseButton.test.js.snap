// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloseButton renders snapshot 1`] = `
<DocumentFragment>
  <button
    class="close"
    title="Dismiss"
  >
    <svg
      class="octicon"
    />
  </button>
</DocumentFragment>
`;

exports[`CloseButton renders with custom title, class name, and other attributes 1`] = `
<DocumentFragment>
  <button
    class="my-class"
    disabled=""
    hidden=""
    title="foofoo"
  >
    <svg
      class="octicon"
    />
  </button>
</DocumentFragment>
`;
